source/actors/player/PlayerHealth.hx

```diff
@@ -0,0 +1,141 @@
+package actors.player;
+
+import flixel.group.FlxGroup;
+
+class PlayerHealth {
+
+    public var healthGraphicGroup(default, null):FlxTypedGroup<HeartContainer>;
+    public var maxHealth(default, null):Int = 12;
+
+    public function new() {
+        initializeGraphics();
+    }
+
+    /**
+	 * Helper function that initializes the heart graphics.
+     * @author Matt Lippelman
+	 */
+    public function initializeGraphics():Void {
+        healthGraphicGroup = new FlxTypedGroup<HeartContainer>();
+        for (val in 0...Math.ceil(maxHealth / 4)) {
+            healthGraphicGroup.add(new HeartContainer(val*32, 0, 4));
+        }
+    }
+
+    /**
+	 * Function to do damage to player.
+     * @author Matt Lippelman
+     * @param damage an integer value in fragments (4 per heart)
+     * @return void
+	 */
+    public function takeDamage(damage:Int):Void {
+        // cycle through hearts from end to beginning
+        var i:Int = healthGraphicGroup.countLiving() - 1;
+        while (i >= 0) {
+            var heart:HeartContainer = healthGraphicGroup.members[i];
+            // test if damage exceeds current value of the heart
+            if (damage > heart.fragments) {
+                // damage exceeds value of heart, damage heart and keep going
+                damage -= heart.fragments;
+                heart.takeDamage(heart.fragments);
+                heart.updateGraphics();
+            } else {
+                // damage is less than value of heart, damage heart and break
+                heart.takeDamage(damage);
+                heart.updateGraphics();
+                break;
+            }
+            i--;
+        }
+
+        if (isDead()) {
+            trace("The player has died");
+        }
```

Comment:
I think we should leave the logical parts of health (whether a player should be dead or not, for example) for the player to track. FlxObject has a built-in health value we can use: https://api.haxeflixel.com/flixel/FlxObject.html#health and it also has overridable methods that trigger when health becomes 0, all of that good stuff we'd want to do logic off of.

Reply:
With this kind of idea in mind, I think it could also make sense to change the `takeDamage` and `heal` methods to be more like a "set" operation than an additive one. Might make the logic simpler, not sure how it fits into the structure you're imagining.